### PR TITLE
MSC3007: Forced insertion and room blocking by self-banning

### PR DESCRIPTION
**Pull request by erkinalp**
This proposal adds forced member insertion, ban reversal without going through unban and join and ability to block rooms by banning self.
[Rendered](https://github.com/matrix-org/matrix-doc/blob/7174e8728c00d4d2e74afed8133c7d3b28532c48/proposals/3007-forced-insertion-and-self-ban.md)

### CODE REVIEW

File: proposals/3007-forced-insertion-and-expiring-membership.md [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+Unless greater than or equal to the power level required to invite, this operation is completely
+disallowed in said room.
+
+Change the auth rules for the member state transistions as follows:
+* from `ban` to `invite`: allowed only if the inviter has power to perform both `ban` and `invite`,
+ill-formed if the user attempts to invite oneself
+* from `ban` to `join`: allowed only if the member attempting to insert has power to perform
+both `ban` and `insert_member`
+* from `leave` to `join`: allowed only if join rule is `public` or the member attempting to insert
+has power to perform `insert_member`
```
Comment (tulir) [inline proposals/3007-forced-insertion-and-expiring-membership.md:28]: This sounds like rooms could arbitrarily force anyone to join, which would be an enormous spam issue

Comment (erkinalp) [reply proposals/3007-forced-insertion-and-expiring-membership.md:28]: Requirement of the power level required to insert being greater than or equal to the power level required to invite greatly solves the problem, albeit not eliminating it.

> Add insert_member field to m.room.power_levels, whose value is going to be interpreted as the minimum power level that a member has to have to be able to forcibly add users to a room. Unless greater than or equal to the power level required to invite, this operation is completely disallowed in said room.

Comment (tulir) [reply proposals/3007-forced-insertion-and-expiring-membership.md:28]: That doesn't solve the problem at all. Anyone can create a room where they're admin.

Comment (erkinalp) [reply proposals/3007-forced-insertion-and-expiring-membership.md:28]: What about allowing self-imposed bans and preventing reinsertion after a user imposes a ban on sefl?

Comment (erkinalp) [reply proposals/3007-forced-insertion-and-expiring-membership.md:32]: Added self-bans as a workaround for rooms that cycle users between banning/kicking and inserting. A self-ban is irreversible and semantically equivalent to a user blocking a room.

File: proposals/3007-forced-insertion-and-expiring-membership.md [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+  "power_level": 0 
+}
+```
+The insertion is ill-formed if the specified power level is greater than user's own power level or
+user attempts to insert itself using this method. Field `roles` is reserved for a future role-based
+power implementation and currently inoperative.
+
+### Expiring invites and memberships
+
+There is currently no way to temporarily insert a member. This seeks to solve the problem by adding
```
Comment (Half-Shot) [inline proposals/3007-forced-insertion-and-expiring-membership.md:54]: State that evaulates differently depending on the system clock could be quite dangerous. You can easily end up with servers that disagree about the current time and therefore would disagree about whether a child event is valid.

File: proposals/3007-forced-insertion-and-expiring-membership.md [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+an `expires` field at invites, which is a timestamp determining the expiration time of both
```
Comment (Half-Shot) [inline proposals/3007-forced-insertion-and-expiring-membership.md:55]: > which is a timestamp determining the expiration time of both the invite and the resulting membership

This means that the join/leave that the client may join/reject the room with also expires? That sounds unexpected and not all that useful. Surely you might want an expiring invite event that allows you to have a persistent join?

File: proposals/3007-forced-insertion-and-expiring-membership.md [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+the invite and the resulting membership. Membership expiration shall be treated as equivalent to
+membership status transtioning to `leave` at the time specified at `expires` field.
```
Comment (Half-Shot) [inline proposals/3007-forced-insertion-and-expiring-membership.md:57]: Because the sender determines the timestamp of an event, I can trivially modify the timestamp of my join/leave event to be before the expiry time long after the event has expired, which makes the expired field pointless. This assumes you are using the origin_server_ts as your timestamp of reference.

If you are using the timestamp at the point at which you got the event from my server, then there is a potential case for disagreement where some servers got my event sooner and believe my join to be valid, and some servers will have got my join late and will believe me to be left.

So if you are using the former system it's trivial to circumvent, and if it's the latter it's trivial to split the room and cause severs to disagree with each other about the state of the room. This proposal cannot work.

File: proposals/3007-forced-insertion-and-expiring-membership.md [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+{
+  "user_id": "@user:example.com",
+  "power_level": 0 
+}
+```
+The insertion is ill-formed if the specified power level is greater than user's own power level or
+user attempts to insert itself using this method. Field `roles` is reserved for a future role-based
+power implementation and currently inoperative.
+
+### Expiring invites and memberships
```
Comment (Half-Shot) [inline proposals/3007-forced-insertion-and-expiring-membership.md:52]: The counter-proposal to this is not to add this logic to room state but have an authoritative bot in the room which tracks membership in the room and rescinds them. This is probably the best way if you want to go off timestamps.

**Review by Half-Shot — CHANGES_REQUESTED** [review]
The expiration system cannot work in a federated environment.

File: proposals/3007-forced-insertion-and-expiring-membership.md (hunk 4 above)
Comment (erkinalp) [reply proposals/3007-forced-insertion-and-expiring-membership.md:57]: There is a middle ground: the originating server checks the expiration when it receives the invite event and then signs the invite with the timestamp according to when it receives. If min(`origin_server_ts`,`origin_client_ts`)>`expires` then reject the invite, else sign the invite as attested. Then creating the joins and kicks as usual for other servers.

**Comment by eras** [comment]
This proposal needs rationale as it can be hard to see why this is needed in the first place.

It should be clarified that in "forcibly add other users into public or knockable rooms" "other users" refers to users currently present in the room (if that is what it means).

**Comment by erkinalp** [comment]
> It should be clarified that in "forcibly add other users into public or knockable rooms" "other users" refers to users currently present in the room (if that is what it means).

It refers to users that *are* present in the room to insert users that are not in the room.

**Comment by eras** [comment]
Hmm, so basically if I create a room and naturally have the `insert_member` permission, I can just pick any people around the whole Matrix and insert them to my room? Wow, is this is the best tool ever to distribute shocking material :thinking:.

I mean it can work in Slack because your employment is on the line for abuse ;), but I cannot see how people would want that feature to Matrix.

If I want it, I can just arrange a client to automatically join any rooms I've been invited to. This would typically be the case as well for schools or companies as well.

**Comment by erkinalp** [comment]
> best tool ever to distribute shocking material 

Protocol-level censorship stopper 😆 

> but I cannot see how people would want that feature to Matrix.

I proposed self-banning as a means to block rooms for that reason.

**Comment by eras** [comment]
One can always create a new room. A whitelist mechanism (a list of acceptable room alias patterns) would be more applicable here and would apply to your use cases. But such a whitelist could easily live in the client.

These use cases are also addressed (better, IMHO) by Spaces #1772 : a company or a school would set up a space and add desired rooms as primary rooms. No, the client still won't be automatically pulled there, but the clients would have the option to do that if programmed to.

**Comment by erkinalp** [comment]
> One can always create a new room. 

It is harder to fork a whole community than to merely create a room, therefore this will happen rarely in practice (except for the deliberate abuse cases that you have mentioned). In those cases, a user could get permission to create rooms revoked by server admin. And so on if the abuser continues doing by creating accounts on other servers.

**Comment by eras** [comment]
Seems like a whole lot of work that could just be basically avoided if the feature wasn't there in the first place.. ?

**Comment by MTRNord** [comment]
I leave this comment just to express that I agree with all the comments asking a) why is this needed b) major spam issue c) WHY? 

TL;DR: PLEASE DONT!

**Comment by erkinalp** [comment]
> a) why is this needed b) major spam issue c) WHY?

a) For school, government and company settings.
b) Yes, you get to abuse, but not continuously (due to the risk of getting your user/server banned from other instances), and it is by design, in order to prove censorship pointless. I laid out possible defence strategies which could be applied by server admins and/or bots. If you do it only in your own server instance, no one can stop you, which kind of fits the design of Matrix, again.
c) Why not?

**Comment by uhoreg** [comment]
In school/government/company settings, usually the organization controls the user's homeserver, and so can add the user to rooms by puppeting the user's account and auto-accepting invites to the rooms.  Aside from that, if the organization does not control the user's homeserver, the current mechanism for inviting users to rooms should be sufficient in most cases.  If it's not sufficient, some way to auto-accept invites would probably work well -- maybe some way of telling the homeserver to automatically accept invites from certain users.  This would not require any new room version (since it doesn't change anything in the room) and it puts control in the target user's hands, as they can say who's invites to automatically accept (which would probably be a short list), rather than waiting for an attacker to attack and reacting to it and having to keep a list of rooms that they don't want to join (which could grow indefinitely).

While I see that this feature could be useful in some situations, I think that in this case the potential for abuse is much greater than the benefits given, especially when there are alternatives that give nearly the same benefits without opening up new attack vectors, and require fewer changes.

**Comment by erkinalp** [comment]
> having to keep a list of rooms that they don't want to join (which could grow indefinitely)

... which would be stored on all parties that participate to that room, by virtue of being implemented as a room ban, meaning ever-growing ban lists mostly comprised of self-bans.
